rfcs/20230324-graph-api-dynamic-shape/README.md

```diff
@@ -0,0 +1,374 @@
+# Proposal for adding dynamic shape support in Graph API
+
+## 1. Background
+
+Dynamic shape is getting more and more attention in deep learning model enabling
+and performance optimization. In contrast to static shape, dynamic shape means
+that the input or output shape of a model or a specific layer is changing from
+iteration to iteration. The typical usages of dynamic shape include:
+
+- The variable sequence length in nature language processing models (e.g.,
+  BERT).
+- The variable mini-batch size in real-time inference serving.
+- The variable spatial size of object detection models in computer vision.
+
+Currently in oneDNN, primitive kernels are usually generated in time (a.k.a.
+JIT) according to src/dst shapes, data types, and other information provided by
+users through primitive creation API. In some cases, kernel generation is time
+consuming, especially on GPU. In order to reduce the kernel generation time,
+oneDNN provides a primitive cache mechanism to cache the kernels and reuse them
+if the same problem is encountered next time. But oneDNN will have to generate
+the kernel again if a different problem shape is provided by users. There are
+two main drawbacks for the primitive cache design:
+
+- It does not help for dynamic shape scenario if the shape changes constantly in
+  a wide range during the execution time. In this case, the primitive cache will
+  probably miss and kernel re-generation will be triggered repeatedly.
+- Even for the static shape scenario, primitive cache does not help for the
+  cases where first-time inference latency is critical for users. For this case,
+  a persistent cache mechanism is exposed to mitigate the issue.
+
+oneDNN provides limited dynamic shape support for few primitives (e.g., matmul
+and reorder). It's achieved by supporting [runtime
+dimensions](https://oneapi-src.github.io/oneDNN/dev_guide_matmul.html#general-notes)
+at primitive creation stage. As mentioned, currently, runtime dimension is not
+supported by all primitives. Even for matmul primitive, once runtime dimension
+is specified, post ops are not fully supported due to [implementation
+limitations](https://github.com/oneapi-src/oneDNN/blob/master/src/common/primitive_attr.cpp#L223).
```

Comment:
nit: better to use permanent link. Same for other code links in this doc.

Reply:
Thanks for reminder. For this, I searched document but seems it's not mentioned explicitly. For others, I changed to use link to document website.